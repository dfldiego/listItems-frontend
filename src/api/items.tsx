import axios from "axios";
import { ITEMS_PATH } from "../config/index";

export const getItemsByQueryAPI = (
  q: string,
  offset?: string,
  limit?: string
) => {
  let url = `${ITEMS_PATH}/items?q=${q}`;
  if (offset) {
    url = `${url}&offset=${offset}`;
  }
  if (limit) {
    url = `${url}&limit=${limit}`;
  }
  return axios.get(url);
};

export const getItemByIdAPI = (id: string) => {
  return axios.get(`${ITEMS_PATH}/items/${id}`);
};
