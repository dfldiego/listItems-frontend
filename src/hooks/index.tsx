export { default as useForm } from "./useForm";
export { default as useIsMounted } from "./useIsMounted";
export { default as useQueryString } from "./useQueryString";
export { default as useItem } from "./useItem";
export { default as useItems } from "./useItems";
