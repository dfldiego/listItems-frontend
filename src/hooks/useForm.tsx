import { useState } from "react";

const useForm = <T extends Object>(initialForm: T) => {
  const [formState, setFormState] = useState(initialForm);

  const onInputChange = (name: string, value: string) => {
    setFormState({
      ...formState,
      [name]: value,
    });
  };

  const onResetForm = () => {
    setFormState(initialForm);
  };

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm,
  };
};

export default useForm;
