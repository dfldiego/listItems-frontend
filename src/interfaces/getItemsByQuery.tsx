export interface GetItemsByQueryResponse {
  data: DataChild;
}

export interface DataChild {
  ok: boolean;
  response: Response;
}

export interface Response {
  categories: any[];
  items: ItemElement[];
  totalPages: number;
}

export interface ItemElement {
  author: Author;
  item: ItemItem;
}

export interface Author {
  nickname: string;
}

export interface ItemItem {
  id: string;
  title: string;
  price: Price;
  picture: string;
  condition: string;
  free_shipping: boolean;
  sellerState: string;
}

export interface Price {
  currency: string;
  amount: number;
  decimals: number;
}
