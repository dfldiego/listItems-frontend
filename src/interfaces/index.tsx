export interface Item {
  author: {
    nickname?: string | null;
  };
  item: {
    id: string;
    title: string;
    price: {
      currency: string;
      amount: number;
      decimals: number;
    };
    picture: string;
    condition: string;
    free_shipping: boolean;
    sold_quantity?: number;
    description?: string;
    sellerState?: string;
  };
}

export interface ProductDetail {
  author: Author;
  item: ItemDetail;
}

export interface Author {
  nickname: null;
}

export interface ItemDetail {
  id: string;
  title: string;
  price: Price;
  picture: string;
  condition: string;
  free_shipping: boolean;
  sold_quantity: number;
  description: string;
}

export interface Price {
  currency: string;
  amount: number;
  decimals: number;
}

export interface ItemsState {
  breadcrumbs: string[] | [];
}
