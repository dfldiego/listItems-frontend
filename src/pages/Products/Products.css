.products_items {
  background-color: var(--white-moderate);
  /* SEGUN DISEÃ‘O ORIGINAL */
  background-color: var(--white);
  grid-column: 3 / 11;
  grid-row: 3 / 11;
}
.button_pagination {
  border-radius: 20px;
  border: 1px solid var(--blue);
  background: var(--blue);
  color: #fff;
  font-weight: bold;
  padding: 12px 45px;
  cursor: pointer;
}
.disabled {
  background: var(--grey);
  border: 1px solid var(--grey);
}
