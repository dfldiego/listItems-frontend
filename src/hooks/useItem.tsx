import { useCallback, useEffect, useState } from "react";

import { getItemByIdAPI } from "../api/items";
import { defaultError } from "../constants";
import { ProductDetail } from "../interfaces";
import useIsMounted from "./useIsMounted";

const useItem = (id: string | undefined) => {
  const isMounted = useIsMounted();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [itemResponse, setItemResponse] = useState<ProductDetail | null>(null);

  const getItem = useCallback(async () => {
    if (!id) return;
    try {
      setLoading(true);
      const response = await getItemByIdAPI(id as string);
      if (isMounted() && response) {
        setItemResponse(response.data.response);
        setError("");
        setLoading(false);
      }
    } catch (error: any) {
      setError(error.response.data.msg ?? defaultError);
      setLoading(false);
    }
  }, [id, isMounted]);

  useEffect(() => {
    getItem();
  }, [getItem]);

  return {
    itemResponse,
    error,
    loading,
  };
};

export default useItem;
