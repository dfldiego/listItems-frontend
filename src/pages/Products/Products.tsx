import { useCallback, useEffect, useState } from "react";
import { getItemsByQueryAPI } from "../../api/items";

import "./Products.css";
import { Item } from "../../interfaces/index";
import useIsMounted from "../../hooks/useIsMounted";
import { defaultError } from "../../constants";

import Breadcrumbs from "../../components/breadcrumbs/Breadcrumbs";
import Search from "../../components/search/Search";
import Loading from "../../components/loading/Loading";
import Error from "../../components/error/Error";
import CardItem from "../../components/cardItem/CardItem";
import { useItems, useQueryString } from "../../hooks";

const Products = () => {
  const isMounted = useIsMounted();
  const { q = "" } = useQueryString();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [items, setItems] = useState<Item[] | never[]>([]);
  const [currentPage, setCurrentPage] = useState(0);
  const [totalPages, setTotalPages] = useState(0);
  const {
    setBreadcrumbs,
    itemsState: { breadcrumbs },
  } = useItems();

  const getProducts = useCallback(async () => {
    if (q === "") return;
    try {
      setLoading(true);
      const response = await getItemsByQueryAPI(
        q as string,
        currentPage.toString(),
        "4"
      );
      if (isMounted() && response) {
        checkRenderBreadcrumbs(response.data.response.categories);
        setItems(response.data.response.items);
        setTotalPages(response.data.response.totalPages);
        setError("");
        setLoading(false);
      }
    } catch (error: any) {
      setError(error.response.data.msg ?? defaultError);
      setLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [q, currentPage, isMounted]);

  useEffect(() => {
    getProducts();
  }, [getProducts]);

  const checkRenderBreadcrumbs = (categoriesResponse: string[] | []) => {
    if (JSON.stringify(breadcrumbs) !== JSON.stringify(categoriesResponse)) {
      setBreadcrumbs(categoriesResponse);
    }
  };

  const handlePrevPage = (newPage: number) => {
    if (currentPage === 0 && newPage < currentPage) return;
    setCurrentPage(newPage);
  };
  const handleNextPage = (newPage: number) => {
    if (currentPage === totalPages && newPage > currentPage) return;
    setCurrentPage(newPage);
  };

  return (
    <div className="page_container">
      <div className="page_search">
        <Search setCurrentPage={setCurrentPage} />
      </div>
      {!error && (
        <div className="page_breadcrumbs">
          <Breadcrumbs />
        </div>
      )}
      {!error && !loading && (
        <>
          <div className="page_content">
            {items.length > 0 &&
              items.map((item) => <CardItem key={item.item.id} item={item} />)}
          </div>
          <div className="pagination_button_one">
            <button
              className={`button_pagination ${
                currentPage === 0 ? "disabled" : ""
              }`}
              onClick={() => handlePrevPage(currentPage - 1)}
              disabled={currentPage === 0}
            >
              &lt;
            </button>
          </div>
          <div className="pagination_button_two">
            <button
              className={`button_pagination ${
                currentPage === totalPages ? "disabled" : ""
              }`}
              onClick={() => handleNextPage(currentPage + 1)}
              disabled={currentPage === totalPages}
            >
              &gt;
            </button>
          </div>
        </>
      )}
      <Loading loading={loading} />
      <Error loading={loading} error={error} />
    </div>
  );
};

export default Products;
