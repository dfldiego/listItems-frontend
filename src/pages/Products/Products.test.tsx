import { render, screen, waitFor } from "@testing-library/react";
import * as router from "react-router-dom";
import * as itemsAPI from "../../api/items";

import { useItems } from "../../hooks";
import { GetItemsByQueryResponse } from "../../interfaces/getItemsByQuery";
import Products from "./Products";

jest.mock("query-string", () => ({
  parse: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
  useNavigate: jest.fn(),
}));

// useItems (mock de custom hook con useContext)
jest.mock("../../hooks/useItems");
export const mockUseItems = useItems as jest.MockedFunction<typeof useItems>;

// useQueryString (mock custom hook)
jest.mock("../../hooks/useQueryString", () => {
  return () => ({
    q: "sarasa",
  });
});

//spy getItemsByQueryAPI
const mockGetItemsByQueryResponse: GetItemsByQueryResponse | any = {
  data: {
    ok: true,
    response: {
      categories: [
        "Arte, Librería y Mercería",
        "Librería",
        "Escolar",
        "Útiles Escolares",
        "Lapiceras",
      ],
      items: [
        {
          author: {
            nickname: "LIBRERIAELPOLI",
          },
          item: {
            id: "MLA1141510338",
            title: "Boligrafos En Gel Pasteles Zebra Sarasa Clip X8 Unidades",
            price: {
              currency: "ARS",
              amount: 7810,
              decimals: 0,
            },
            picture:
              "http://http2.mlstatic.com/D_769302-MLA50238468858_062022-O.jpg",
            condition: "new",
            free_shipping: false,
            sellerState: "Capital Federal",
          },
        },
        {
          author: {
            nickname: "LIBRERIAELPOLI",
          },
          item: {
            id: "MLA1141516866",
            title:
              "Boligrafos En Gel Metalizados Zebra Sarasa Clip X9 Unidades",
            price: {
              currency: "ARS",
              amount: 9020,
              decimals: 0,
            },
            picture:
              "http://http2.mlstatic.com/D_934921-MLA50238473277_062022-O.jpg",
            condition: "new",
            free_shipping: true,
            sellerState: "Capital Federal",
          },
        },
        {
          author: {
            nickname: "ELBAZARDIGITAL",
          },
          item: {
            id: "MLA772013592",
            title: "10x Zebra Sarasa Clip 0.3 Mm Retractable Gel Ink Pen",
            price: {
              currency: "ARS",
              amount: 15229,
              decimals: 0,
            },
            picture:
              "http://http2.mlstatic.com/D_887555-MLA29460710543_022019-O.jpg",
            condition: "new",
            free_shipping: true,
            sellerState: "Buenos Aires",
          },
        },
        {
          author: {
            nickname: "ELBAZARDIGITAL",
          },
          item: {
            id: "MLA752494135",
            title: "Set X 5 Zebra Sarasa Clip 0.5mm Ballpoint Pen",
            price: {
              currency: "ARS",
              amount: 9699,
              decimals: 0,
            },
            picture:
              "http://http2.mlstatic.com/D_640131-MLA31600382812_072019-O.jpg",
            condition: "new",
            free_shipping: true,
            sellerState: "Buenos Aires",
          },
        },
      ],
      totalPages: 16,
    },
  },
};

describe("<Products>", () => {
  const navigate = jest.fn();
  beforeEach(() => {
    mockUseItems.mockReturnValue({
      setBreadcrumbs: jest.fn(),
      itemsState: {
        breadcrumbs: [],
      },
    });
    jest.spyOn(router, "useNavigate").mockImplementation(() => navigate);
  });
  test("should render correctly", async () => {
    const spyGetItemsByQuery = jest
      .spyOn(itemsAPI, "getItemsByQueryAPI")
      .mockReturnValue(mockGetItemsByQueryResponse);
    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
    await waitFor(() => {
      expect(spyGetItemsByQuery).toBeDefined();
    });
  });

  test("should contain search component", () => {
    render(<Products />);
    const searchComponent = screen.getByTestId("search-component-test-id");
    expect(searchComponent).toBeInTheDocument();
  });

  test("should contain breadcrumbs component", () => {
    render(<Products />);
    const searchComponent = screen.getByTestId("breadcrumbs-component-test-id");
    expect(searchComponent).toBeInTheDocument();
  });
});
