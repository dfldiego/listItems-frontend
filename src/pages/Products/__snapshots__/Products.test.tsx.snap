// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products> should render correctly 1`] = `
<div>
  <div
    class="page_container"
  >
    <div
      class="page_search"
    >
      <div
        class="search_container"
        data-testid="search-component-test-id"
      >
        <div
          class="search_logo"
        >
          <img
            alt="MercadoLibre_logo"
            class="search_logo_img"
            data-testid="meli-logo-test-id"
            src="Logo_ML.png"
          />
        </div>
        <div
          class="search_form"
        >
          <form
            aria-label="form"
            class="search_form_row"
          >
            <input
              autocomplete="off"
              class="input-text"
              name="searchText"
              placeholder="Nunca dejes de buscar"
              type="text"
              value="sarasa"
            />
            <button
              class="search_button"
              type="submit"
            >
              <div>
                <img
                  alt="Logotipo Buen Sabor"
                  class="search_img"
                  src="ic_Search.png"
                />
              </div>
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="page_breadcrumbs"
    >
      <ul
        class="breadcrumbs_ul"
        data-testid="breadcrumbs-component-test-id"
      />
    </div>
    <div
      class="spinner mt5 loadingGrid"
      data-testid="loading-test-id"
    />
  </div>
</div>
`;
